nvflare/private/fed/server/job_cmds.py

```diff
@@ -656,6 +697,51 @@ def download_job(self, conn: Connection, args: List[str]):
                 self._clean_up_download(conn, tx_id)
                 conn.append_error("internal error", meta=make_meta(MetaStatusValue.INTERNAL_ERROR))
 
+    def download_job_components(self, conn: Connection, args: List[str]):
+        """Download additional job components (e.g., ERRORLOG_site-1) for a specified job.
+
+        Based on job download but downloads the additional components for a job that job download does
+        not download.
+        """
+        job_id = args[1]
+        engine = conn.app_ctx
+        job_def_manager = engine.job_def_manager
+        if not isinstance(job_def_manager, JobDefManagerSpec):
+            self.logger.error(
+                f"job_def_manager in engine is not of type JobDefManagerSpec, but got {type(job_def_manager)}"
+            )
+            conn.append_error("internal error", meta=make_meta(MetaStatusValue.INTERNAL_ERROR))
+            return
+
+        # It is possible that the same job is downloaded in multiple sessions at the same time.
+        # To allow this, we use a separate sub-folder in the download_dir for each download.
+        # This sub-folder is named with a transaction ID (tx_id), which is a UUID.
+        # The folder path for download the job is: <download_dir>/<tx_id>/<job_id>.
+        tx_id = str(uuid.uuid4())  # generate a new tx_id
+        job_download_dir = self.tx_path(conn, tx_id)  # absolute path of the job download dir.
+        with engine.new_context() as fl_ctx:
+            try:
+                list_of_data = job_def_manager.list_client_data(jid=job_id, fl_ctx=fl_ctx)
+                if list_of_data:
+                    job_components = [
+                        item for item in list_of_data if item not in {"workspace", "meta", "scheduled", "data"}
+                    ]
+                for job_component in job_components:
+                    job_def_manager.get_storage_for_download(
+                        job_id + "_components", job_download_dir, job_component, job_component, fl_ctx
+                    )
+                self.download_folder(
+                    conn,
+                    tx_id=tx_id,
+                    folder_name=job_id + "_components",
+                    download_file_cmd_name=AdminCommandNames.DOWNLOAD_JOB_FILE,
+                )
+            except Exception as e:
+                secure_log_traceback()
+                self.logger.error(f"exception downloading job {job_id}: {secure_format_exception(e)}")
+                self._clean_up_download(conn, tx_id)
+                conn.append_error("internal error", meta=make_meta(MetaStatusValue.INTERNAL_ERROR))
```

Comment:
The "download_job" method also has some duplicate codes as this part, maybe we could separate out the common downloading/checking logic to another method and reuse it.

Reply:
we can clean this up in a future PR